### ISSUE TITLE
[DFA-arch T2] Selective-scan / Mamba-lite + DFA feedback — the TOP NOVELTY pick (greenfield; kill BPTT's seq-length memory)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#155]
**Tier 2 — top novelty.** GREENFIELD: no DFA/forward-forward/DRTP on Mamba/S4/S5 exists (only vanilla-RNN precedent: RFLO, Murray eLife 2019; e-prop, Bellec Nat.Comms 2020). Best structural fit in the whole survey — the SSM core is a LINEAR RECURRENCE, and a random-feedback error injected per step composes through the same linear operator, so random-feedback credit assignment is mathematically natural here in a way it is not for attention. Concrete value: kill BPTT/scan backward memory (scales with seq-len; Mamba's kernel already recomputes states), enable online/streaming training.

## Ask
A small **selective-scan / linear-recurrence SSM** block (S4-lite or Mamba-lite: input-dependent A,B,C selection + gating + a conv branch), stackable, with a SMALL-OUTPUT-DIM head (sequence classification / regression / reconstruction — NOT an LM head). Per-layer `--policy chain|dfa`, DFA feedback injected at the block/step level.

## Success target + the sharp caveat
all-DFA-SSM matches BP on a small-output seq task at k-times-less activation memory. **CRITICAL:** in the purely-LINEAR case FA collapses to plain gradient descent, so the interesting alignment MUST live in the NONLINEAR parts (selection A,B,C / gating / conv). The experiment must isolate that — a control with the nonlinearity removed should show DFA==GD (no alignment story), and the DFA-vs-GD gap in the nonlinear model is the finding.

## Consuming Tao experiment: F19 (SSM), the novel flag.
Refs: RFLO eLife 2019; e-prop Nat.Comms 2020. No DFA-SSM precedent (that's the point).

**Comment by OriPekelman** [comment]
Shipped in `773f5eb` — **the success target is met at one cut and the ticket's structural argument comes out backwards, which is the more useful half.**

`toy train ssm` (runner `libexec/toy-train-ssm`, engine `lib/toy/llm/engine/ssm_engine.rb`, task `lib/toy/io/toy_ssm_task.rb`, gate `prep/ssm_gate.rb`): a stack of channel-wise **selective** linear recurrences — input-dependent decay + causal depthwise conv + gating — with a small-output sequence-classification head reading the **last** timestep, per-layer `--policy chain|dfa|frozen`, and a `--dfa-cut layer|step` axis. CPU-only (tao#18/#19).

**The recurrence is unrolled from differentiable primitives, and that is not a shortcut.** ggml ships fused SSM_SCAN/SSM_CONV and the shim exposes them, but `ggml_compute_backward` covers 43 ops and neither is among them — inference-only, they abort under autodiff. It is also the honest construction: the BPTT graph the DFA arm claims to avoid has to exist before avoiding it means anything.

## 1. The result

Delayed-cue task (one marked cue in the first quarter, noise elsewhere, readout at step 64), 600 steps, matched init + seed:

| selection | seed | BP | **DFA (layer cut)** | DFA (step cut) | frozen |
|---|---|---|---|---|---|
| selective | 0 | 1.000 | **0.996** | 0.250 | 0.227 |
| selective | 1 | 0.992 | **0.988** | 0.250 | 0.227 |
| selective | 2 | 0.988 | **1.000** | 0.238 | 0.238 |
| lti | 0 | 0.750 | 0.645 | 0.645 | 0.305 |
| lti | 1 | 0.730 | 0.621 | 0.668 | 0.270 |
| lti | 2 | 0.738 | 0.656 | 0.617 | 0.266 |

**The mandatory bar is MET at the layer cut**: DFA tracks BP to within .012 and clears the frozen control by .76. That is the program's third positive, on the architecture with no DFA precedent at all.

## 2. But the ticket's thesis inverts

This ticket argues: *"the SSM core is a LINEAR RECURRENCE, and a random-feedback error injected per step composes through the same linear operator, so random-feedback credit assignment is mathematically natural here."* And it warns: *"in the purely-LINEAR case FA collapses to plain gradient descent, so the interesting alignment MUST live in the NONLINEAR parts."*

Read the DFA(step) column against that. **Cutting BPTT entirely is FREE in the linear model** (step .645 vs layer .645, equal to within noise) **and CATASTROPHIC in the selective one** (.99 → chance). Not an LR artifact: swept 3e-5 … 1e-2, the best per-step cell is .355 against a .227 frozen control and a 1.000 BP.

So the linearity that makes random feedback compose neatly is exactly what makes the per-step cut cost nothing — and the selection that makes the model *good* is exactly what makes it fail. **The kill-BPTT value prop is available only in the regime where the model cannot do the task.**

Both halves are gated, because either alone is misreadable: the lti equality alone looks like "the step cut works", the selective collapse alone looks like a broken build. Together they are the finding — and the lti equality is also the proof that the step cut is correctly *wired*, since a dead cut could not learn at all.

## 3. The activation-memory target does NOT hold in this harness

The stated success target is "matches BP at k-times-less activation memory". Measured (realized graph nodes; every arm linear in T):

| T | BP | DFA (layer) | DFA (step) |
|---|---|---|---|
| 16 | 814 | 858 | 1077 |
| 32 | 1646 | 1722 | 2165 |
| 64 | 3310 | 3450 | 4341 |
| 128 | 6638 | 6906 | 8693 |

The step cut's graph is **31% bigger**, not smaller — the detach dups and the per-step surrogate terms are net additions. In a graph-based autodiff every forward tensor is materialised whatever the credit rule, so the streaming memory win is not expressible as a graph rewrite; it needs a streaming implementation. `cost.graph_nodes` carries that caveat in the bundle so nobody reads the number as bytes saved.

**This is worth deciding on before anyone spends the deferred CUDA twin** (tao#19's exception): a device port would not show the memory win either.

## 4. Two construction findings

1. **A per-step linear surrogate needs a tap that can change the prediction.** My first build tapped the layer output at every step. With a last-step readout, the *final* layer's `o_t` for t < T−1 has no functional path to the prediction at all — so the global error is not a proxy target there, it is not a target at all, the linear surrogate has nothing to balance it, and the weights run away (measured: loss 1e24, then NaN). The state `h_t` is the only per-step quantity that always has such a path, because it is what propagates forward in time. General rule for injecting DFA into any recurrent or late-readout architecture.
2. **The unreachable-input landmine fired again** (same shape as toy#154's scalar-error label and franken-moe's `t_hp` under `--optimizer sgd`): under the layer cut there are no state taps, so that family's feedback matrix is consumed by nothing, gets no backend buffer, and the upload aborts inside `ggml_backend_tensor_set`. Allocate a B only when its tap family is non-empty.

## 5. Running it

```
toy train ssm --steps 600 --seed 0 --policy chain,chain
toy train ssm --steps 600 --seed 0 --policy dfa,dfa                 # layer cut
toy train ssm --steps 600 --seed 0 --policy dfa,dfa --dfa-cut step
toy train ssm --steps 600 --seed 0 --policy frozen,frozen
#  ... and the whole grid again under --selection lti
```

Arms differ ONLY in `--policy` / `--dfa-cut` / `--selection`. `--dfa-cut` is deliberately not `--dfa-granularity`: that one picks matmul-vs-block in *depth*, this picks layer-vs-step in *time*.

**`--align-events` is rejected on this lane**, and that is structural rather than an omission: the DFA update arrives through autodiff from the surrogate roots, so it lands in the *same accumulator* a BP run would use and there is no second tensor to take a cosine against. toy#158 hit this first on macro-DFA. The gate uses "the B seed moves the curve, and is inert on chain" instead — the assertion that actually catches a silently-unwired build.

`--task mean` is this lane's `blobs`: the class signal is spread over every step, so neither memory nor selection is needed and a frozen recurrence already integrates it. Selectable, measured, asserted.

## 6. What is not there

The **long-sequence CUDA twin** (tao#19's deferred exception) is not built, and per §3 it would not demonstrate the memory claim anyway. The **eligibility-trace family** (RFLO / e-prop), which is the principled way to give per-step credit under a late readout, is not implemented either — §4.1 is essentially the reason it exists, and it would be the natural follow-up if F19 wants the per-step cut to actually work.

**Comment by OriPekelman** [comment]
Closing: shipped in `773f5eb` — feat(dfa-arch): toy#155/F19 selective-scan lane — and the ticket's thesis, INVERTED

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
